docs: detail oauth configuration options and common cases (#2217)
slackapi/bolt-js: docs/content/basic/authenticating-oauth.md

slug: /concepts/authenticating-oauth
---

> OAuth allows installation of your app to any workspace and is an important step
> in distributing your app. For each completed authentication process, you receive
> unique [access tokens and related information](#the-installation-object) that
> can be retrieved for incoming events and used to make scoped API requests.

Reviewer: ```suggestion
OAuth allows installation of your app to any workspace and is an important step in distributing your app. For each completed authentication process, you receive unique [access tokens and related information](#the-installation-object) that can be retrieved for incoming events and used to make scoped API requests.
```

docusaurus quirk: The unfurl will show the first _line_  so for the first paragraph best to keep it on one line to stop it only showing half a sentence.

Author: Shoot I will forget this often... But TIL! I'm wondering if this could be checked by a linter at some point? Like, the first non-frontmatter line can only be a single line? IDK. I **can** hope to remember this for now though!

I feel like it'd be nice to ask ya- are broken lines elsewhere something you're a fan of? I find it easier to edit, but please do not let that bias an answer 🙏

> All of the additional underlying details around authentications can be found at

Reviewer: ```suggestion
All of the additional underlying details around authentications can be found within
```

> [api.slack.com/authentication/oauth-v2][oauth-v2]!

Reviewer: ```suggestion
[the Slack API documentation][oauth-v2]!
```

## Configuring the application

To set your Slack app up for distribution, you will need to enable Bolt OAuth
and store installation information securely. Bolt supports OAuth by using

> [`@slack/oauth`][oauth-node] to handle most of the work; this includes setting

Reviewer: ```suggestion
the [`@slack/oauth`][oauth-node] package to handle most of the work; this includes setting
```

or... "library"? not sure what is proper. But we like to say what the thing is instead of just saying the thing. Example: instead of saying "use `conversations.info` to do it" it should be "use the `conversations.info` method to do it"

Author: 🧠 Excellent practice here- I'm in quick agreement! Thanks for finding these fixes 🙏

I'll keep `eye` on myself for this one 😅

> up OAuth routes, state verification, and passing your app an installation object

Reviewer: ```suggestion
up OAuth routes, verifying state, and passing your app an installation object
```

uhhh would this make sense maybe? I was trying to match the verbs. But it might not make sense

Author: TIL the verb-noun-verb sandwhich isn't digested well 😔

which you must store.

### App options

The following `App` options are required for OAuth installations:

> - `clientId`: unique identifier of the application client.

Reviewer 2: The source code says "Client ID, which can be found under the Basic Information section of your application on https://api.slack.com/apps". Something like this would be more helpful for users.

Author: This is way more helpful 👏 Updated with something similar!

- `clientSecret`: secret value to confirm the client ID.

> - `stateSecret`: secret value used for [state verificiation][verification] of

Reviewer 2: you can use better wording tho
```suggestion
- `stateSecret`: secret value used for state parameter string generation plus [its verificiation][verification] of
```

Author: 🤓 Learning a bit more with each comment, thank you for these! Added with similar wording 👍

>   authentication requests.

Reviewer 2: nit: 
```suggestion
  authorization requests.
```

  information.

#### Development and testing

Here we've provided a default implementation of the `installationStore` with
[`FileInstallationStore`][store-file-installation] which can be useful when
developing and testing your app:

```javascript
const { App } = require("@slack/bolt");
const { FileInstallationStore } = require("@slack/oauth");
const app = new App({
  signingSecret: process.env.SLACK_SIGNING_SECRET,
  clientId: process.env.SLACK_CLIENT_ID,
  clientSecret: process.env.SLACK_CLIENT_SECRET,
  stateSecret: "my-state-secret",
  scopes: ["channels:history", "chat:write", "commands"],
  installationStore: new FileInstallationStore(),
});
```

:::warning

This is **not** recommended for use in production - you should
[implement your own installation store](#installation-store). Please continue
reading or inspect [our example apps][examples].

:::

### Installer options

We provide several options for customizing default OAuth using the
`installerOptions` object, which can be passed in during the initialization of
`App`. You can override these common options and
[find others here][install-provider-options]:

> - `authVersion`: Settings for either new Slack Apps (`v2`) or Classic Slack Apps

Reviewer 2: We can say most developers need to set "v2" anyway. To clarify things, we can say "v1" was available for "classic" slack app, which is a very old version of Slack app model, but it's no longer available for newly created apps.

Author: 🤯 This makes a fast read so much easier on the mind

  (`v1`). Default: `v2`.

> - `directInstall`: Skip the [default installation page][installation] at

Reviewer 2: skip **rendering (or displaying)** the page?

Author: Added "rendering" here but also a few more words about the alternative 🪙 🤔  Still open to more updates or removing a bit!

  `installPath`. Default: `false`.
- `installPath`: Path of the URL for starting an installation. Default:
  `/slack/install`.

> - `metadata`: Relevant session information passed between requests. Optional.

Reviewer 2: unfortunately, this value is static per app. thus, developers cannot use it for enhancing request data (that's why bolt-python/bolt-java didn't implement this and we haven't received any requests to add). it's okay to mention this but supported use cases are limited

Author: TIL- this is interesting but I agree that use cases are limited if state is available 😉

- `redirectUriPath`: Path of the installation callback URL. Default:
  `/slack/oauth_redirect`.

> - `stateVerification`: Option to skip state verification for requests. Default:

Reviewer 2: how about having type as well? it would make this part even easier to understand
```suggestion
- `stateVerification: boolean`: Option to skip state verification for requests. Default:
```

Author: Awesome callout! I made a small update to formatting, but this makes this reference so much better 🙏

  `true`.
- `userScopes`: User scopes to request during installation. Default: `[]`.
- `callbackOptions`: Customized [responses to send][callbacks] during OAuth.
  Default: [`CallbackOptions`][callback-options-default].

> - `stateStore`: Replace the `stateSecret` with a [custom state store][state].

Reviewer 2: This is a module, which generates and verifies state parameter. Thus, I don't think this description explains what it is well.

We can say:
Customizable OAuth state parameter handling module; ClearStateStore should work for most scenarios. However, if you need even better security, implementing your own module that stores the state parameter data on the server-side database would be a good approach.

Author: Good to know we can recommend `ClearStateStore` with confidence! It reads much better when this is treated as a true module too, thanks for catching this!

is also optional and can be skipped using `installerOptions.directInstall`.

Inspect this [example app][direct-install] and snippet below:

```javascript
const app = new App({
  signingSecret: process.env.SLACK_SIGNING_SECRET,
  // ...
  installerOptions: {
    // highlight-start
    directInstall: true,
    installPath: "/slack/installations", // www.example.com/slack/installations
    // highlight-end
  },
});
```

#### Add to Slack button

The [default][add-to-slack] `Add to Slack` button initiates the OAuth process
with Slack using a generated installation URL. If customizations are wanted to
this page, changes can be made using
[`installerOptions.renderHtmlForInstallPath`][installation-page] and the
generated installation URL:

```javascript
const app = new App({
  signingSecret: process.env.SLACK_SIGNING_SECRET,
  // ...
  installerOptions: {
    // highlight-start
    renderHtmlForInstallPath: (addToSlackUrl) => {
      return `<a href="${addToSlackUrl}">Add to Slack</a>`;
    },
    // highlight-end
  },
});
```

We do recommend using the provided [button generator][oauth-v2] when formatting
links to the authorization page!

:::note

> Authorization requests with changed or additional scops require

Reviewer: ```suggestion
Authorization requests with changed or additional scopes require
```

Author: Sad because I actually needed `scops` to land in this PR 🪐 🔭

[generating a unique authorization URL](#extra-authorizations).

:::

### Redirect URL

Bolt for JavaScript provides the **Redirect URL** path `/slack/oauth_redirect`
out-of-the-box for Slack to use when redirecting users that complete the OAuth
installation flow.

You will need to add the full **Redirect URL** including your app domain in
[app settings][settings] under **OAuth and Permissions**, e.g.
`https://example.com/slack/oauth_redirect`.

To supply a custom Redirect URL, you can set `redirectUri` in the App options
and `installerOptions.redirectUriPath`. Both must be supplied and be consistent
with the full URL if a custom Redirect URL is provided:

```javascript
const app = new App({
  signingSecret: process.env.SLACK_SIGNING_SECRET,
  clientId: process.env.SLACK_CLIENT_ID,
  clientSecret: process.env.SLACK_CLIENT_SECRET,
  stateSecret: "my-state-secret",
  scopes: ["chat:write"],

>   // highlight-next-line

Reviewer: docusaurus experttttttt

Author: I'm in my pterodactylus era ✨ 🐦

  redirectUri: "https://example.com/slack/redirect",
  installerOptions: {
    // highlight-next-line
    redirectUriPath: "/slack/redirect",
  },
});
```

#### Custom callbacks

The page shown after OAuth is complete can be changed with
`installerOptions.callbackOptions` to display different details:

```javascript
const app = new App({
  signingSecret: process.env.SLACK_SIGNING_SECRET,
  // ...
  installerOptions: {
    // highlight-start
    callbackOptions: {
      success: (installation, installOptions, req, res) => {
        res.send("The installation succeeded!");
      },
      failure: (error, installOptions, req, res) => {
        res.send("Something strange happened...");
      },
    },
    // highlight-end
  },
});
```

Full reference reveals [these additional options][callback-options] but if no
options are provided, the [`defaultCallbackSuccess`][callback-default-success]

> and [`defaultCallbackFailure`][callback-default-failure] are used.

Reviewer: hmmm i want to make these have nouns attached to them but not sure what is right. "callback"?

`fetchInstallation`, and `deleteInstallation` for each part of this process. The
following implements a simple installation store in memory, but persistent
storage is strongly recommended for production:

```javascript
const app = new App({
  signingSecret: process.env.SLACK_SIGNING_SECRET,
  clientId: process.env.SLACK_CLIENT_ID,
  clientSecret: process.env.SLACK_CLIENT_SECRET,
  stateSecret: "my-state-secret",
  scopes: ["chat:write", "commands"],
  installationStore: {
    storeInstallation: async (installation) => {
      if (installation.team !== undefined) {
        return await database.set(installation.team.id, installation);
      }
      throw new Error("Failed to save installation data to installationStore");
    },
    fetchInstallation: async (installQuery) => {
      if (installQuery.teamId !== undefined) {
        return await database.get(installQuery.teamId);
      }
      throw new Error("Failed to fetch installation");
    },
    deleteInstallation: async (installQuery) => {
      if (installQuery.teamId !== undefined) {
        return await database.delete(installQuery.teamId);
      }
      throw new Error("Failed to delete installation");
    },
  },
});
```

> Lookups for `fetchInstallation` happen as part of the built-in

Reviewer: ```suggestion
Lookups for the `fetchInstallation` handler happen as part of the built-in
```

[`authorization`][authorization] of incoming events and provides app listeners

> with the `context.botToken` for convenient use.

Reviewer: ```suggestion
with the `context.botToken` object for convenient use.
```

    this.store[key] = value;
  },
};
```

</details>

---

## Additional cases

The above sections set your app up for collecting a bot token on workspace
installations with handfuls of configuration, but other cases might still be
explored.

### User tokens

User tokens represent workspace members and can be used to
[take action on behalf of users][user-tokens]. Requesting user scopes during
installation is required for these tokens to be issued:

```javascript
const app = new App({
  signingSecret: process.env.SLACK_SIGNING_SECRET,
  clientId: process.env.SLACK_CLIENT_ID,
  clientSecret: process.env.SLACK_CLIENT_SECRET,
  scopes: ["chat:write", "channels:history"],
  installerOptions: {
    userScopes: ["chat:write"],
  },
});
```

Most OAuth processes remain the same, but the
[`installation`](#the-installation-object) object received in
`storeInstallation` has a `user` attribute that should be stored too:

```javascript
{
  team: { id: "T012345678", name: "example-team-name" },
  user: {
    token: "xoxp-314159-26*********-********************",
    scopes: ["chat:write"],
    id: "U01234567"
  },
  tokenType: "bot",
  appId: "A01234567",
  // ...
}
```

> Successful `fetchInstallation` lookups will also include the `context.userToken`

Reviewer: ```suggestion
Successful `fetchInstallation` lookups will also include the `context.userToken` object
```

associated with the received event in the app listener arguments.

:::note

> The `tokenType` remains `"bot"` while `scopes` are requested, even with the

Reviewer: ```suggestion
The `tokenType` value remains `"bot"` while `scopes` are requested, even with the
```

```javascript
const app = new App({
  signingSecret: process.env.SLACK_SIGNING_SECRET,
  clientId: process.env.SLACK_CLIENT_ID,
  clientSecret: process.env.SLACK_CLIENT_SECRET,
  scopes: ["chat:write"],
  installerOptions: {
    // highlight-next-line
    stateVerification: false,
  },
});
```

To learn more about the OAuth installation flow with org-wide apps,
[read the API documentation][org-ready-oauth].

#### Org-wide installation objects

Being installed to an organization can grant your app access to multiple
workspaces and the associated events.

##### The org-wide `installation` object

The `installation` object from installations to a team in an organization have
an additional `enterprise` object and `isEnterpriseInstall` set to either `true`
or `false`:

```javascript
{
  team: undefined,
  enterprise: { id: "E0000000001", name: "laboratories" },
  user: { token: undefined, scopes: undefined, id: "U0000000001" },
  tokenType: "bot",
  isEnterpriseInstall: true,
  appId: "A0000000001",
  authVersion: "v2",
  bot: {
    scopes: [
      "chat:write",
    ],
    token: "xoxb-000001-00*********-********************",
    userId: "U0000000002",
    id: "B0000000001"
  }
}
```

> Apps installed org-wide will receive `isEnterpriseInstall` as `true`, but apps

Reviewer: ```suggestion
Apps installed org-wide will receive the `isEnterpriseInstall` parameter as `true`, but apps
```

could also still be installed to individual workspaces in organizations. These

> apps receive installation information for both the `team` and `enterprise`:

Reviewer: ```suggestion
apps receive installation information for both the `team` and `enterprise` parameters:
```

```javascript
{
  team: { id: "T0000000001", name: "experimental-sandbox" },
  enterprise: { id: "E0000000001", name: "laboratories" },
  // ...
  isEnterpriseInstall: false,
  // ...
}
```

##### The org-wide `installQuery` object

This `installQuery` object provided to the `fetchInstallation` and

> `deleteInstallation` handlers is familiar but with an `enterpriseId` defined and

Reviewer: ```suggestion
`deleteInstallation` handlers is the same as ever, but now with an additional value, `enterpriseId`, defined and
```

        return await database.delete(installQuery.teamId);
      }
      throw new Error("Failed to delete installation");
    },
  },
  // highlight-end
});
```

<details>
  <summary>Example database object</summary>

For quick testing purposes, the following might be interesting:

```javascript
const database = {
  store: {},
  async get(key) {
    return this.store[key];
  },
  async delete(key) {
    delete this.store[key];
  },
  async set(key, value) {
    this.store[key] = value;
  },
};
```

</details>

### Sign in with Slack

Right now Bolt does not support [Sign in with Slack][siws] out-of-the-box. This
still continues to remain an option using APIs from the
[`@slack/web-api`][web-api] package, which aim to make implementing OpenID
Connect (OIDC) connections simple. Alternative [routes][custom-routes] might be
required.

Explore [this relevant package documentation][oidc] for reference and example.

### Extra authorizations

If you need additional authorizations, such as user tokens from users of a team
where your app is already installed, or have a reason to dynamically generate an
install URL, an additional installation is required.

> Generating a new installation URL requires a few steps: manually instantiate an
> `ExpressReceiver`, assign the instance to a variable named `receiver`, and then
> call `receiver.installer.generateInstallUrl()`.

Reviewer: ```suggestion
Generating a new installation URL requires a few steps: 
1. Manually instantiate an `ExpressReceiver` instance.
2. Assign the instance to a variable named `receiver`.
3. Call the `receiver.installer.generateInstallUrl()` function.
```

Author: Didn't realize why listing these steps inline felt so wrong! 🤪  Thanks for this suggetsion 🙏

Read more about `generateInstallUrl()` in the
[OAuth docs][generate-install-url].

### Common errors

Occasional mishaps in various places throughout the OAuth process can cause
errors, but these often have meaning! Explore [the API documentation][errors]
for additional details for common error codes.

> [add-to-slack]: https://github.com/slackapi/node-slack-sdk/blob/main/packages/oauth/src/default-render-html-for-install-path.ts

Reviewer: i had no idea you could do this! 🥳

Author: It is such an important feature. Blessed markdown 🕴️

});
```

:::warning

This is **not** recommended for use in production - you should
[implement your own installation store](#installation-store). Please continue
reading or inspect [our example apps][examples].

:::

### Installer options

We provide several options for customizing default OAuth using the
`installerOptions` object, which can be passed in during the initialization of
`App`. You can override these common options and
[find others here][install-provider-options]:

- `authVersion`: `string`. Settings for either new Slack apps (`v2`) or
  "classic" Slack apps (`v1`). Most apps use `v2` since `v1` was available for a
  Slack app model that can no longer be created. Default: `v2`.
- `directInstall`: `boolean`. Skip rendering the
  [installation page](#add-to-slack-button) at `installPath` and redirect to the
  authorization URL instead. Default: `false`.
- `installPath`: `string`. Path of the URL for starting an installation.
  Default: `/slack/install`.
- `metadata`: `string`. Static information shared between requests as install
  URL options. Optional.
- `redirectUriPath`: `string`. Path of the installation callback URL. Default:
  `/slack/oauth_redirect`.

> - `stateVerification`: `boolean`. Option to skip state verification for

Author: Oh woah, great catch! That even confused me 😭  I made an update using a lot of the same wording, but please let me know if more adjustments are wanted!

Thanks once more for the thorough readings and review 🙏
